src/engine.rs

```diff
@@ -864,18 +900,21 @@ impl Reedline {
                                 if let Some(binding) = self.find_keybinding(modifiers, code) {
                                     self.run_edit_commands(&binding);
                                 }
+                                self.line_buffer.reset_olds();
                                 return Ok(Signal::CtrlC);
                             }
                             (KeyModifiers::CONTROL, KeyCode::Char('l'), EditMode::Emacs) => {
                                 self.tab_handler.reset_index();
+                                self.line_buffer.reset_olds();
                                 return Ok(Signal::CtrlL);
                             }
                             (KeyModifiers::NONE, KeyCode::Char(c), x)
                             | (KeyModifiers::SHIFT, KeyCode::Char(c), x)
                                 if x == EditMode::ViNormal =>
                             {
                                 self.tab_handler.reset_index();
-                                self.run_edit_commands(&[EditCommand::ViCommandFragment(c)]);
+                                self.run_edit_commands(&[ViCommandFragment(c)]);
+                                self.line_buffer.set_previous_lines();
```

Comment:
For vi mode the undo behavior will probably have to take additional pecularities into account, that are proabably outside the scope of this PR